### ISSUE TITLE
Applications plain client_secret

### ISSUE THREAD

**Issue by AlonFasta** [evonove/django-oauth-toolkit#276]
Hi,

Today Applications client secret saved in db as plain text.
To improve confidentiality and security its possible to use django.auth.contrib.hashers to hash and salt applications client secret.

The security advantages are obvious - even DB read access or DB dump file wont let intruders to use our  client secrets to gain access tokens to the eco system resources.

The downside is that the service clients will have to save and manage their secrets separately from the Django admin / applications back office page - over there secrets will be shown hashed and salted.

Backwards compatibility: Running with old style - plain client secrets will be possible due to new migration that will hash and salt existing secrets.

I have already forked the repo to implement it, do you have interest in that feature anyway?

Thanks,
Alon Fast

**Comment by masci** [comment]
Premise: I'm pretty far from being a security expert :)

I think the responsibility to secure the authorization server against this specific threat model should be up to the server's maintainers. Sure we could do it, but if we need to secure DOT against a violation of the database server, we should also encrypt refresh tokens and access tokens, maybe authorization tokens (grants) too. And, at that point, where should we stop?

@synasius any consideration on this matter?

**Comment by AlonFasta** [comment]
Well i think that tokens are pretty different case, 
each token has its expiration date, more than that every consumer of this app can configure its own expiration time to its token and manage them to be expired after every request or revoking them programatically if needed.
Client secret can be used to generate those tokens and impersonate the owner of the token.

I agree that server's maintainers do have responsibility to secure the authorization server against this specific threat model, but it comes as different layer of security.
My perspective - To create real layered security both of the above should happen.

Thanks,
Alon

**Comment by attiasr** [comment]
+1

**Comment by roievt** [comment]
+1

**Comment by jleclanche** [comment]
Because the Application model can be overridden, I'm going to go ahead and close this as a wontfix. You can implement your own app model and add in any security layers you like, which is far better than DOT taking these decisions on its own imho.

**Comment by lordgordon** [comment]
Hi there,

I understand the reasons to set this bug as wontfix. Anyway I want to spend my two cents :)

It is undeniable that every user should do its best to protect the db containing the client secret, and that it's possible to implement a custom app model with the desired security level.

Yet, nowadays it's considered bad practice to store secret values in plain text. Encrypted secrets allows to reduce damages in case the database is violated. Furthermore, it should not be possible by database admins to view client's secrets. Every big player I use doesn't even allow to view again a client secret once it is created (Google, Bitbucket, DigitalOcean, GitHub, ...).

For those reasons I think django-oauth-toolkit should provide this additional layer by default, as it cheap to implement and greatly decrease the surface attack.

Unfortunately I currently don't have time to study a PR for this repository. Yet I'll gladly publish my custom application model as soon it is ready for the better of django-oauth-toolkit users.

Thanks for your attention.

Best Regards,

Nicholas

**Comment by jleclanche** [comment]
@lordgordon the problem is that i don't want DOT to be dictating the way those secrets are stored. I'll take suggestions but unless they're readily one-size-fits-all they're unlikely to see a merge.

**Comment by lordgordon** [comment]
@jleclanche I would suggest this:

- the default setting is the current "no encryption" application model (to keep this module backward compatible);
- a "secure" application model storing encrypted secrets with the same methods Django uses for user's password.

Then, if an user has very specific security levels should implement its own application model.

This way DOT is well balanced between security out-of-the-box, aligned with the default Django security, and usability for advanced customizations.

**Comment by jleclanche** [comment]
User passwords are not encrypted, they are hashed. This is unsuitable for client secrets.

Django does not encrypt anything.

**Comment by lordgordon** [comment]
My bad, I was using the term "encryption" in its broader meaning. 

Of course the technical implementation to store passwords, therefore client secrets, it's an hash-based algorithm, such HMAC for PBKDF2 (Django's default), or KDF as it is implemented in Argon2, another common and modern option to safely store secrets one-way.

**Comment by jleclanche** [comment]
You can't hash client secrets. Their actual value is necessary.

**Comment by lordgordon** [comment]
I see.  If I correclty understood your code, the problem lies at https://github.com/evonove/django-oauth-toolkit/blob/master/oauth2_provider/oauth2_validators.py#L91 where ```client_secret``` is supposed to be plain text.

So it is not possible to implement what me and OP were asking with just a custom application model.

**Comment by attiasr** [comment]
@lordgordon @jleclanche I don't agree, I see no difference between user passwords and client secrets, the can validated in the same way (hash the secret and validate it agains the stored hash..)
it is generally a bad practice to store secrets or password in clear text in a database.

**Comment by jleclanche** [comment]
I have my doubts. Github, for example, who are known to be religious about security, do not hash client secrets.

This is an impractical decision to take imho and protocol-wise I'm not even sure it's feasible (I don't know oauth2 well enough but the value of the secret may be required in-transit). If it does turn out to be feasible, we could store both a client_secret and client_secret_hash, with a setting to not store client_secret at all and only rely on the latter (or some similar behaviour which would ensure the client_secret is never stored on disk).

@attiasr if you want to work on a PR for such behaviour be my guest, but i recommend looking at whether this is possible in the protocol itself beforehand.

**Comment by lordgordon** [comment]
@jleclanche your claim about GitHub is interesting. Could you please share your source?

Anyway, this is not a matter of "personal taste". Secrets must be secured, it's a best practice nowadays. I already explained before the scenarios.

Furthermore, I think you are completely misunderstanding how  "hashing the client secret" (or a password, as it is the same) works. It is  not a matter of protocol (OAuth 2) as the client secret is not hashed client side but server side! The client secret is, of course, sent to the server "as it is" (no hashing, hopefully secured by a TSL connection). Then as soon the server receives the secret, it is hashed and the hash is compared with the one on the database.

I hope is clear now what is the issue here.

Best Regards,

Nicholas

**Comment by jleclanche** [comment]
@lordgordon create an app and see for yourself: https://github.com/settings/developers

I know full well how hashing works, but my faint memory of oauth2 is that the client secret's actual value is required *on the server side as part of the protocol itself* as its value is used to calculate responses. Like I said I'm not certain on this but if it's the case, hashing is a no go. If it's only used informationally then yes we can go for it, but double check first.

**Comment by lordgordon** [comment]
@jleclanche That's your "source"? Afterall, once you create a new GitHub's token you cannot see it again (*). This strongly suggests they store it hashed!

I think you are missing the whole point of this thread: our request is to hash the client_secret in the database. Yet the user see the plain text client_secret only once when the new application is created.

In fact, in the original [OAuth2 RFC](https://tools.ietf.org/html/rfc6749) there is no mention of 
the client_secret being used to "calculate responses". The client_secret is only used to authenticate a client ("Client Authentication", section 2.3) and is explicitly called "client password" (section 2.3.1), to further clarify!

I double-checked within this repository's code, and oauthlib as well, to see if, for some weird reason, such value is used other than for authentication purposes. I was unable to find any reference other than client's auth.

## I strongly suggest to reopen this bug.
This will help to have a PR and to keep visibility on this issue.

(*): 
![image](https://cloud.githubusercontent.com/assets/7745293/24043323/6c81aab2-0b16-11e7-9c33-52989d1990c1.png)

**Comment by jleclanche** [comment]
Yes it's my source. Here's a now-deleted example app if that's not clear enough:

![image](https://cloud.githubusercontent.com/assets/235410/24043792/04399a54-0b21-11e7-8cae-8704b369cc7f.png)

Looking around, Dropbox also seems to do the same. In fact it seems pretty common that the client secrets are not hashed. They may be encrypted of course but I'm taking a pretty strong stance here that we should not dictate encryption requirements for users of the oauth2_provider library as they differ from user to user.

Like i said. if someone wants to work on a hashing setting, I'm fine with this as long as it's optional and they actually make sure it works.

**Comment by vandersonmota** [comment]
I arrived late at the party, but still:
>For each registered application, you’ll need to store the public client_id and the private client_secret. Because these are essentially equivalent to a username and password, you should not store the secret in plain text, instead only store an encrypted or hashed version, to help reduce the likelihood of the secret leaking.

https://www.oauth.com/oauth2-servers/client-registration/client-id-secret/
